ddtrace/contrib/cherrypy/middleware.py

```diff
@@ -0,0 +1,114 @@
+"""
+Datadog trace code for cherrypy.
+"""
+
+# stdlib
+import logging
+
+# project
+from ddtrace import config
+from .. import trace_utils
+from ... import compat
+from ...ext import errors, SpanTypes
+from ...propagation.http import HTTPPropagator
+
+# 3p
+import cherrypy
+from cherrypy.lib.httputil import valid_status
+
+log = logging.getLogger(__name__)
+
+SPAN_NAME = "cherrypy.request"
+
+
+class TraceTool(cherrypy.Tool):
+    def __init__(self, app, tracer, service, use_distributed_tracing):
+        self.app = app
+        self._tracer = tracer
+        self._service = service
+        self._use_distributed_tracing = use_distributed_tracing
+
+        cherrypy.Tool.__init__(self, "on_start_resource", self._on_start_resource, priority=95)
+
+    def _setup(self):
+        cherrypy.Tool._setup(self)
+        cherrypy.request.hooks.attach("on_end_request", self._on_end_request, priority=5)
+        cherrypy.request.hooks.attach("after_error_response", self._after_error_response, priority=5)
+
+    def _on_start_resource(self):
+        if self._use_distributed_tracing:
+            propagator = HTTPPropagator()
+            context = propagator.extract(cherrypy.request.headers)
+            # Only need to active the new context if something was propagated
+            if context.trace_id:
+                self._tracer.context_provider.activate(context)
+
+        try:
+            cherrypy.request._datadog_span = self._tracer.trace(
+                SPAN_NAME,
+                service=self._service,
+                span_type=SpanTypes.WEB,
+            )
+        except Exception:
+            log.debug("cherrypy: error tracing request", exc_info=True)
+
+    def _after_error_response(self):
+        error_type = cherrypy._cperror._exc_info()[0]
+        error_msg = str(cherrypy._cperror._exc_info()[1])
+        error_tb = cherrypy._cperror.format_exc()
+
+        self.close_span(error_type, error_msg, error_tb)
+
+    def _on_end_request(self):
+        self.close_span()
+
+    def close_span(self, error_type=None, error_msg=None, error_tb=None):
+        span = getattr(cherrypy.request, "_datadog_span", None)
+
+        if not span or not span.sampled:
+            return
+
+        if error_type:
+            span.error = 1
+            span.set_tag(errors.ERROR_TYPE, error_type)
+            span.set_tag(errors.ERROR_MSG, error_msg)
+            span.set_tag(errors.STACK, error_tb)
+        else:
+            span.error = 0
+
+        # Let users specify their own resource in middleware if they so desire.
+        # See case https://github.com/DataDog/dd-trace-py/issues/353
+        if span.resource == SPAN_NAME:
+            path_info = cherrypy.request.path_info
+            span.resource = compat.to_unicode(path_info).lower()
```

Comment:
For the resource name we typically try to do the pattern `{method} {path}` (eg. `GET /path/to/resource`). Eg. in flask https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/flask/patch.py#L270-L274

However we want the resource to aggregate as much as possible. Eg. we'd want `GET /users/1`, `GET /users/2`, `GET /users/3`, ... to be aggregated as `GET /users` for the resource.

I'm not familiar enough with cherrypy yet to have any idea how it resolves urls to handlers, but we'd want to try to re-use that logic if possible.

Please let me know if this doesn't make any sense, resource names are a bit tricky!

Reply:
Yeah, this is a tough one. I've now followed the style of `{method} {path}`.

I also had a go at substituting path variables with their variable name, which would have ended up looking like `GET /users/{{id}}`. We can get the path variable keys and values in a params dict of the `cherrypy.request`, however, I didn't feel comfortable substituting the URL as that involved guesswork. I found in CherryPy's documentation a [neat tool for exactly this job](https://docs.cherrypy.org/en/latest/_modules/cherrypy/_cpdispatch.html#Dispatcher.find_handler) - but it didn't behave as I expected, so I have left a couple of notes and moved on.

Resource aggregation based on path variables may need to be a future feature until we can dig into this more, if that is acceptable to you?

Reply:
Having the variable included is just as good (if not better!) so this is great 😄

Reply:
Yeah, variables would work nicely in the case of more complex resource structures, such as the classic `GET /album/{{album_name}}/track/{{track_name}}`. But I hit a dead end with this one. If it's acceptable to you, I'd like to come back to this in a future PR.